core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java

```diff
@@ -496,7 +496,14 @@ private static Consumer<BuildChainBuilder> loadStepsFromClass(Class<?> clazz,
                         final Class<? extends SimpleBuildItem> buildItemClass = parameterClass
                                 .asSubclass(SimpleBuildItem.class);
                         methodStepConfig = methodStepConfig.andThen(bsb -> bsb.consumes(buildItemClass));
-                        methodParamFns.add((bc, bri) -> bc.consume(buildItemClass));
+                        methodParamFns.add((bc, bri) -> {
+                            SimpleBuildItem bi = bc.consume(buildItemClass);
+                            if (bi == null) {
+                                throw reportError(parameter,
```

Comment:
A better place for this check would be to check in `io.quarkus.builder.BuildContext#doProduce` on the producing side rather than on the consuming side. Then the build step and the item ID can also be logged.

Reply:
The thing is that it's ok to produce `null` if all consumers inject `Optional`. Do we have this information inside the `BuildContext#doProduce()`?

Reply:
> The thing is that it's ok to produce null if all consumers inject Optional

Do we have such cases?

Reply:
This is a documentation error and should also be corrected.

For example (I did not write this):

> If there was no `/config.xml` resource available in the build environment, then a `null` `RuntimeServiceBuildItem` would be returned and no subsequent logic based on a `RuntimeServiceBuildItem` being produced would execute.

However it also says (IIRC I wrote this):

> Normally it is an error for a step which is included to consume a simple build item that is not produced by any other step. In this way, it is guaranteed that all the declared values will be present and non-`null` when a step is run.

The latter is correct and logically excludes the former. We promise that all build items are non-`null`; we also break that promise. But there is no better time to fix it than now.

Reply:
Ok. I think you're right. Except that I've seen many use cases like:

```java
@BuildStep
SomeBuildItem produce(...) {
  if (someCondition) {
     return null;
  }
  return new SomeBuildItem();
}
```
BTW the docs also states: _"Multi build values are always considered optional. If not present, an empty list will be injected."_. Does that mean that returning `null` for a `MultiBuildItem` is ok? From the code it looks like we just add `null` to the resulting list?

Reply:
> Except that I've seen many use cases like

Theoretically we could easily convert these to `BuildProducer`

Reply:
My point is that if we support `Optional<>` even for produced `null` simple build items then we could avoid NPEs and break less build steps. Moreover, we could do the same for multi build items (i.e. ignore `null` values).

Reply:
> > Except that I've seen many use cases like
> 
> Theoretically we could easily convert these to `BuildProducer`

Hm, define "easily convert". I think that it's going to be a lot of manual work..

Reply:
Of course it's manual work :)

Reply:
Breaking in this case is good. The more unintended behavior you allow, the more unintended consequences you sign up for. Things may be allowed today that are not OK, and this is one of them. We explicitly should not allow `null`. I'd break things now to avoid bugs later.

Reply:
> Ok. I think you're right. Except that I've seen many use cases like:
> 
> ```java
> @BuildStep
> SomeBuildItem produce(...) {
>   if (someCondition) {
>      return null;
>   }
>   return new SomeBuildItem();
> }
> ```
> 
> BTW the docs also states: _"Multi build values are always considered optional. If not present, an empty list will be injected."_. Does that mean that returning `null` for a `MultiBuildItem` is ok? From the code it looks like we just add `null` to the resulting list?

No, this would also be a problem. We need to ensure that the lists are all `null`-free. Additionally to the above reasons, having lists containing `null` doesn't interoperate well with `List.of(...)` which rejects `null` entries.

Reply:
> Of course it's manual work :)

A LOT OF manual work :D

Reply:
Are you sure? Have you seen many cases where we return `null`?

Reply:
> Breaking in this case is good. The more unintended behavior you allow, the more unintended consequences you sign up for. Things may be allowed today that are not OK, and this is one of them. We explicitly should not allow `null`. I'd break things now to avoid bugs later.

I don't say it's bad ;-). But from my point of view we don't need to be that strict and still avoid unintended behavior. In any case, it seems that this topic would deserve more attention. We should probably send an email to the quarkus-dev.

> No, this would also be a problem. We need to ensure that the lists are all null-free. Additionally to the above reasons, having lists containing null doesn't interoperate well with List.of(...) which rejects null entries.

I meant conceptually. Since the values are considered optional. But ofc we would need to either fail or ignore `null`s.

Reply:
> Are you sure? Have you seen many cases where we return `null`?

Like I said take a look at the examples in this PR ;-).

https://github.com/quarkusio/quarkus/blame/main/core/deployment/src/main/java/io/quarkus/deployment/ide/IdeProcessor.java#L82C2-L82C2

https://github.com/quarkusio/quarkus/blame/main/core/deployment/src/main/java/io/quarkus/deployment/console/ConsoleProcessor.java#L129-L131

Reply:
Ah, sorry. It's not. It's the consumed `EffectiveIdeBuildItem`. It should have been: https://github.com/quarkusio/quarkus/blame/main/core/deployment/src/main/java/io/quarkus/deployment/ide/IdeProcessor.java#L134-L136

And I do remember that I used it as well several times.

Reply:
Right, I am sure there are a few, but I my guess is that there are not too many

Reply:
So while looking at the code it seems that it's not a problem of a build step method returning `null` (which is [just ignored here](https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java#L683-L686)) but in general, if there's no build item produced then `null` is injected - the `ExtensionLoader` just calls `io.quarkus.builder.BuildContext.consume(Class<T>)` which returns `null` if no single result exists. In other words, we have the same problem if a build step declares a `BuildProducer<MyItem> myItemProducer` param but produces nothing or if `myItemProducer.produce(null)` is used.

The same applies to `MultiBuildItem`.

That changes the situation a little bit. What's interesting - the javadoc of `io.quarkus.builder.BuildContext.consume(Class<T>)` states: `Returns: the produced item (may be null)`. So I'm a bit lost to be honest and I need to step back and think about the consequences again :thinking:.

Reply:
I think `null` might be consumable in the raw-consume case because that class is not aware of `Optional`, and `null` is used internally for this condition. In other words, if the dependency is optional, then `null` is returned here. However if we block `null` from being produced, then this is the *only* case where the raw `consume` method could return `null`, which is consistent.

Reply:
> Therefore if we forbid null from being produced, then everything else will already be correct.

No, not really. It does not prevent unexpected NPE. The following build steps demonstrate the problem I described above: https://github.com/mkouba/ext-test/blob/main/deployment/src/main/java/org/acme/ext/test/deployment/ExtTestProcessor.java. `ExtTestProcessor#produceMyBuildItem()` does not produce anything and so `ExtTestProcessor#feature()` receives `null`.

Reply:
Producing nothing should be treated identically to producing `null`: it is an error.

Reply:
Hm, I used this pattern many times and it's even documented in some ArC build items, for example `io.quarkus.arc.deployment.ValidationPhaseBuildItem`.

This definitely deserves a wider community discussion. I will send an email to the quarkus-dev next Monday...

Reply:
> Producing nothing should be treated identically to producing null: it is an error.

Do you mean if `BuildProducer<WhateverBuildIte> producer` is declared and not invoked, it should be an error?
If so, we can't realistically do this, it would break a ton of stuff

Reply:
It should be an error - this only works by accident. If that means we need to take a bit longer to fix it then so be it IMO. Again I'm happy to take this on if @mkouba doesn't want to tackle this.

Reply:
So IIUC, you want to move to a state where there is no conditional production of build items?
I don't think that's a realistic goal at this point

Reply:
It can be conditional, but the condition should be checked before the build chain runs (using a predicate for example). Normally though steps should be written differently (see chat discussion; this is a longer term goal though).

Reply:
Yeah, I understand the goal for everything to be declarative, but my point is that I am very skeptical we can do it since there is a lot of code written like this.